akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala

```diff
@@ -500,30 +502,43 @@ private[persistence] trait Eventsourced extends Snapshotter with PersistenceStas
    *
    * All incoming messages are stashed.
    */
-  private def recovering(recoveryBehavior: Receive) = new State {
-    override def toString: String = "replay started"
-    override def recoveryRunning: Boolean = true
+  private def recovering(recoveryBehavior: Receive) = {
+    // protect against replay stalling forever because of journal overloaded and such
+    context.setReceiveTimeout(extension.settings.recovery.replayEventTimeout)
```

Comment:
somewhere you should also remove/restore the receive timeout when the recovery has completed